### PR TITLE
feat(stats): Emit outcomes for applied rate limits

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Emit outcomes to represent events and attachments being removed from an
envelope by project rate limits. The main difference is in
`EnvelopeLimiter`, which returns a tuple of `Enforcement` and
`RateLimits` used for emitting outcomes:

- Enforcements declare the quantities of categories that have been rate
  limited with the individual reason codes that caused rate limiting. If
  multiple rate limits applied to a category, then the longest limit is
  reported.
- Rate limits declare all active rate limits, regardless of whether they
  have been applied to items in the envelope.
- Rate limits for sessions are not reported.

# Example

**Interaction between Events and Attachments**

An envelope with an `Error` event and an `Attachment`. Two quotas
specify to drop all attachments (reason `"a"`) and all errors
(reason `"e"`). The result of enforcement will be:

1. All items are removed from the envelope.
2. Enforcements report both the event and the attachment dropped with
   reason `"e"`, since dropping an event automatically drops all
   attachments with the same reason.
3. Rate limits report the single event limit `"e"`, since attachment
   limits do not need to be checked in this case.

**Required Attachments**

An envelope with a single Minidump `Attachment`, and a single quota
specifying to drop all attachments with reason `"a"`:

1. Since the minidump creates an event and is required for processing,
   it remains in the envelope and is marked as `rate_limited`.
2. Enforcements report the attachment dropped with reason `"a"`.
3. Rate limits are empty since it is allowed to send required
   attachments even when rate limited.

**Previously Rate Limited Attachments**

An envelope with a single item marked as `rate_limited`, and a quota
specifying to drop everything with reason `"d"`:

1. The item remains in the envelope.
2. Enforcements are empty. Rate limiting has occurred at an earlier
   stage in the pipeline.
3. Rate limits are empty.

### CODE REVIEW

File: relay-server/src/endpoints/common.rs [hunk 1]
```diff
@@ -405,134 +405,147 @@ where
...
-                        None => {
-                            // if we don't have a trace context we can't do dynamic sampling so stop.
-                            return Box::new(Ok((envelope, rate_limits, None)).into_future())
-                                as RetVal;
+        .and_then(clone!(
+            event_id,
+            envelope_summary,
+            scoping,
+            outcome_producer,
+            |project| {
```
Comment (RyanSkonnord) [inline relay-server/src/endpoints/common.rs]: 😱 Rustfmt sure had some opinions about this change. "Hide whitespace changes" recommended, I guess.

File: relay-server/src/actors/project.rs [hunk 2]
```diff
@@ -595,6 +605,60 @@ impl Actor for Project {
...
+            envelope_summary: EnvelopeSummary::compute(&envelope),
+            outcome_producer: message.outcome_producer.clone(),
+            scoping: *scoping,
+        }
+    }
+
+    fn emit_rate_limit_outcomes(&self, applied_limits: &RateLimits) {
+        for applied_limit in applied_limits.iter() {
+            if applied_limit.categories.is_empty() {
+                // Empty categories value indicates that the rate limit applies to all data.
```
Comment (RyanSkonnord) [inline relay-server/src/actors/project.rs]: Per https://github.com/getsentry/relay/blob/master/relay-quotas/src/rate_limit.rs#L140-L141. So far I have only blind faith in that comment to support that this is correct/necessary behavior, but maybe the integration tests will clarify things as I continue digging.

**Comment by RyanSkonnord** [comment]
Revised to-do list:

- [X] Move outcome producer from `CheckEnvelope` to `Project` (this removes the earlier problem with the `Debug` trait)
- [X] Add envelope summary update to `common.rs`
- [x] Add envelope summary update to `events.rs`
- [x] Fix failing integration tests afterward, if any
- [ ] Add new integration test coverage, if needed (on that weird "empty categories" case?)

The current integration test failures are because there are redundant outcomes in the event category. ~~I'm guessing this is because they aren't being removed from the envelope summary in `events.rs` after rate-limiting. I think there should be no failures once that's fixed.~~ [Update: Nope. [See below](https://github.com/getsentry/relay/pull/951#discussion_r600132886).]

File: relay-server/src/actors/events.rs [hunk 3]
```diff
@@ -1594,7 +1593,12 @@ impl Handler<HandleEnvelope> for EventManager {
                 }
 
                 match processed.envelope {
-                    Some(envelope) => Ok(envelope),
+                    Some(envelope) => {
+                        // TODO: Fix scope problem and uncomment
+                        // envelope_summary.replace(EnvelopeSummary::compute(&envelope));
```
Comment (RyanSkonnord) [inline relay-server/src/actors/events.rs]: ~~@jan-auer Can you advise on this? I'm not sure whether I need to propagate `envelope_summary` through all the `and_then` closures up to this point, enclose them all in one big parent closure (which is what makes this simpler in `common.rs`, AFAICT), or something else.~~

Update: Never mind, resolved in be73b2bd. That turned out to be embarrassingly simple. 🤦 (In my defense, I had to take another pass at wrapping my head around what `clone!` does in order to understand why that works. Though I'm a little surprised I hadn't tried the same thing by accident already.)

File: relay-server/src/actors/events.rs [hunk 4]
```diff
@@ -1594,8 +1593,14 @@ impl Handler<HandleEnvelope> for EventManager {
...
-                    Some(envelope) => Ok(envelope),
-                    None => Err(ProcessingError::RateLimited(rate_limits)),
+                    Some(envelope) => {
+                        envelope_summary.replace(EnvelopeSummary::compute(&envelope));
+                        Ok(envelope)
+                    }
+                    None => {
+                        envelope_summary.replace(EnvelopeSummary::empty());
+                        Err(ProcessingError::RateLimited(rate_limits))
+                    }
```
Comment (RyanSkonnord) [inline relay-server/src/actors/events.rs]: I believe at least one integration test is failing because a rate-limited item is still in the envelope here, making the outcome emitted below redundant to the new one. That means that updating the envelope summary is essentially a no-op. From debug logs, it looks like the item *is* being correctly removed by `RateLimit::enforce`, so I'm not clear why a non-empty item list would in the `CheckedEnvelope`. The order of the debug logs also implies that `RateLimit::enforce` is happening *after* this code section, so I seem to be misunderstanding something here.

[Update: [This may have changed something](https://github.com/getsentry/relay/pull/951/commits/168b6e400efdb49551d634749171fc3b3831a8ec) but I think the tests are still failing for the same reason.)

Comment (jan-auer) [reply relay-server/src/actors/events.rs]: There is indeed one case in which we retain rate limited items. Minidump attachments are _both_ errors and attachments. It works like this:

- Assume an organization has run out of attachment quota, we will need to reject all attachments now.
- A minidump comes in. Since the organization still has errors quota, we need to process the minidump.
- We check the rate limiter and it tells us to drop the attachment. Now we take two actions:
    1. We emit an outcome for the dropped attachment.
    2. Leave the minidump item in but mark it as "rate_limited" in its header.
- In all subsequent rate limiting checks, the rate limited item is ignored since we already emitted an outcome.
- After processing, we store the processed event but drop the attachment without another outcome.

`EnvelopeSummary` already has a check to compensate for that and ignores items with the `rate_limited` header set. However, we might have missed a case there.

File: relay-server/src/actors/project.rs [hunk 5]
```diff
@@ -600,6 +616,63 @@ impl Actor for Project {
     }
 }
 
+struct RateLimitEnvelope {
+    event_id: Option<EventId>,
+    remote_addr: Option<IpAddr>,
```
Comment (jan-auer) [inline relay-server/src/actors/project.rs]: You could move these two into `EnvelopeSummary`. It looks like this type is a utility that would fit well into `utils::rate_limits`

File: relay-server/src/actors/project.rs (hunk 2 above)
Comment (jan-auer) [reply relay-server/src/actors/project.rs]: I believe this approach is problematic since you cannot reconstruct the accurate drop reason from the `RateLimits` structure, unfortunately. This is an inherent flaw of `RateLimits` and we can consider changing that, too. 

To illustrate, consider the following example:
- There is a single quota `category:error limit:0`. It drops all error events.
- `EnvelopeLimiter` will also drop all attachments in the same envelope as a result of that.
- When you check `RateLimits` here, you won't find the `attachment` category, even though you just dropped them. 

From the top of my head, I have two ideas to solve that:
1. Move emitting outcomes into `EnvelopeLimiter`, because that's where you can keep track of the dropped quantities.
2. From `EnvelopeLimiter::enforce`, return an instance of `EnvelopeSummary` that contains the dropped quantities. Then, use that summary (+ scoping) to emit outcomes. This approach is preferable as it decouples concerns.

File: relay-server/src/actors/project.rs [hunk 6]
```diff
@@ -685,13 +690,14 @@ impl Project {
             Ok(self.rate_limits.check_with_quotas(quotas, item_scoping))
         });
 
-        let rate_limits = envelope_limiter.enforce(&mut envelope, scoping)?;
+        let rate_limits = envelope_limiter.enforce(&mut envelope, scoping, |outcome| {
+            self.outcome_producer.do_send(outcome)
+        })?;
```
Comment (RyanSkonnord) [inline relay-server/src/actors/project.rs]: I'd like to be able to just pass `self.outcome_producer` to `enforce` rather than this awkward closure. I did it this way to accommodate the `EnvelopeLimiter` unit tests, so that we can use a simple closure as a mock in place of the outcome producer. Let me know if you can suggest a better way to mock out the outcome producer.

The goal of injecting the outcome producer into `enforce` is to obviate the `RateLimitEnforcement` struct, as in d4af165760091cce83af67c3a5a3a7c905cb597c. I'm not certain it's worth it. If it isn't, we can just revert that commit.

[EDIT: Never mind, it broke some stuff I hadn't noticed. I've reverted it. The aforementioned commit is still in the history if you want to explore my idea but I don't think it's a high priority at all.]

File: tests/integration/test_outcome.py [hunk 7]
```diff
@@ -439,40 +439,40 @@ def _get_message(message_type):
         raise Exception("Invalid message_type")
 
 
-@pytest.mark.parametrize("category_type", ["session", "transaction"])
-def test_no_outcomes_rate_limit(
+@pytest.mark.parametrize("category_type", [("session", False), ("transaction", True)])
+def test_outcomes_rate_limit(
     relay_with_processing, mini_sentry, outcomes_consumer, category_type
 ):
     """
```
Comment (untitaker) [inline tests/integration/test_outcome.py]:
```suggestion
@pytest.mark.parametrize("category,is_outcome_expected", [("session", False), ("transaction", True)])
def test_outcomes_rate_limit(
    relay_with_processing, mini_sentry, outcomes_consumer, category, is_outcome_expected
):
    """
```

File: tests/integration/test_outcome.py [hunk 8]
```diff
@@ -439,40 +439,40 @@ def _get_message(message_type):
...
+    Tests that outcomes are emitted or not, depending on the type of message.
 
-    Pass a transaction that is rate limited and check that an outcome is not emitted (although
-    the transaction does NOT go through).
-
-    NOTE: This test should start failing once transactions outcomes become supported.
-    Once that happens change test to verify that a transaction outcome IS sent
+    Pass a transaction that is rate limited and check whether a rate limit outcome is emitted.
     """
+    category, is_outcome_expected = category_type
```
Comment (untitaker) [inline tests/integration/test_outcome.py]:
```suggestion
    """
```

**Review by untitaker — APPROVED** [review]
